src/libsystemd/sd-bus/bus-dump.c

```diff
@@ -502,57 +505,90 @@ int bus_creds_dump(sd_bus_creds *c, FILE *f, bool terse) {
-        uint32_t incl_len;       /* number of octets of packet saved in file */
-        uint32_t orig_len;       /* actual length of packet */
-} pcaprec_hdr_t;
+static uint16_t pcapng_optlen(size_t len) {
+        return ALIGN4(len + sizeof(struct pcapng_option));
+}
 
-int bus_pcap_header(size_t snaplen, FILE *f) {
+static void pcapng_opt_write(FILE *f, uint16_t code, const void *data, size_t len) {
+        _cleanup_free_ struct pcapng_option *opt = NULL;
+        uint32_t optlen = ALIGN4(sizeof(*opt) + len);
+
+        opt = calloc(1, optlen);
+        opt->code = code;
+        opt->length = len;
+        if (len > 0)
+                memcpy(opt->data, data, len);
+        fwrite(opt, 1, optlen, f);
+}
 
-        pcap_hdr_t hdr = {
-                .magic_number = 0xa1b2c3d4U,
-                .version_major = 2,
-                .version_minor = 4,
-                .thiszone = 0, /* UTC */
-                .sigfigs = 0,
-                .network = 231, /* D-Bus */
+static void pcapng_section_header(FILE *f, const char *os, const char *app) {
+        uint32_t len;
+
+        /* determine length of section header and options */
+        len = sizeof(struct pcapng_section_header);
+        len += pcapng_optlen(strlen(os));
+        len += pcapng_optlen(strlen(app));
+        len += pcapng_optlen(0);	/* OPT_END */
+        len += sizeof(uint32_t);	/* trailer length */
+
+        struct pcapng_section_header hdr = {
+                .block_type = PCAPNG_SECTION_BLOCK,
+                .block_length = len,
+                .byte_order_magic = PCAPNG_BYTE_ORDER_MAGIC,
+                .major_version = PCAPNG_MAJOR_VERS,
+                .minor_version = PCAPNG_MINOR_VERS,
+                .section_length = UINT64_MAX,
         };
 
-        if (!f)
-                f = stdout;
+        fwrite(&hdr, 1, sizeof(hdr), f);
+        pcapng_opt_write(f, PCAPNG_SHB_OS, os, strlen(os));
+        pcapng_opt_write(f, PCAPNG_SHB_USERAPPL, app, strlen(app));
+        pcapng_opt_write(f, PCAPNG_OPT_END, NULL, 0);
+        fwrite(&len, 1, sizeof(uint32_t), f);
+}
+
+/* Only have a single instance of dbus pseudo interface */
+static void pcapng_interface_block(FILE *f, uint32_t snaplen) {
+        uint32_t len;
 
         assert(snaplen > 0);
         assert((size_t) (uint32_t) snaplen == snaplen);
 
-        hdr.snaplen = (uint32_t) snaplen;
+        /* no options (yet) */
+        len = sizeof(struct pcapng_interface_block) + sizeof(uint32_t);
+        struct pcapng_interface_block hdr = {
+                .block_type = PCAPNG_INTERFACE_BLOCK,
+                .block_length = len,
+                .link_type  = 231, /* D-Bus */
+                .snap_len = snaplen,
+        };
 
         fwrite(&hdr, 1, sizeof(hdr), f);
+        fwrite(&len, 1, sizeof(uint32_t), f);
+}
 
+int bus_pcap_header(size_t snaplen, FILE *f) {
+        struct utsname u;
+        const char appinfo[]
+                = "systemd " STRINGIFY(PROJECT_VERSION) " (" GIT_VERSION ")";
```

Comment:
This creates an array on the stack. Why not just use this string directory in the argument?

Reply:
sure will do